clients/spark/core/src/main/scala/io/treeverse/clients/ApiClient.scala

```diff
@@ -198,28 +221,70 @@ class ApiClient private (conf: APIConfigurations) {
    *  translate that URL to use an appropriate Hadoop FileSystem.
    */
   def getMetaRangeURL(repoName: String, commitID: String): String = {
-    val commit = commitsApi.getCommit(repoName, commitID)
+    val commit = getCommit(repoName, commitID)
     getMetaRangeURL(repoName, commit)
   }
 
   /** Query lakeFS for a URL to the range of rangeID of repoName and
    *  translate that URL to use an appropriate Hadoop FileSystem.
    */
   def getRangeURL(repoName: String, rangeID: String): String = {
-    val range = metadataApi.getRange(repoName, rangeID)
+    val getRange = new dev.failsafe.function.CheckedSupplier[StorageURI]() {
+      def get(): StorageURI = metadataApi.getRange(repoName, rangeID)
+    }
+    val range = retryWrapper.wrapWithRetry(getRange)
     val location = range.getLocation
     URI
       .create(getStorageNamespace(repoName, StorageClientType.HadoopFS) + "/" + location)
       .normalize()
       .toString
   }
 
-  def getBranchHEADCommit(repoName: String, branch: String): String =
-    branchesApi.getBranch(repoName, branch).getCommitId
+  def getBranchHEADCommit(repoName: String, branch: String): String = {
+    val getBranch = new dev.failsafe.function.CheckedSupplier[Ref]() {
+      def get(): Ref = branchesApi.getBranch(repoName, branch)
+    }
+    val response = retryWrapper.wrapWithRetry(getBranch)
+    response.getCommitId
+  }
 
   // Instances of case classes are compared by structure and not by reference https://docs.scala-lang.org/tour/case-classes.html.
   case class StorageNamespaceCacheKey(
       repoName: String,
       storageClientType: StorageClientType
   )
 }
+
+class RequestRetryWrapper(
+    val readTimeout: Int,
+    val maxDurationSeconds: Double = -1,
+    val maxNumRetries: Int = 5
+) {
+  val UnsetMaxDuration = -1
+
+  var maxDuration = maxDurationSeconds
+  if (maxDuration == UnsetMaxDuration) {
+    maxDuration = maxNumRetries * readTimeout + 7.1 * 1
+  }
+
+  private val retryPolicy: Policy[Any] = RetryPolicy
+    .builder()
+    .withBackoff(1, 20, ChronoUnit.SECONDS)
+    .withJitter(.25)
+    .withMaxRetries(maxNumRetries)
+    .withMaxDuration(Duration.ofSeconds(maxDuration.asInstanceOf[Number].longValue))
```

Comment:
Why not
```suggestion
    .withMaxDuration(Duration.ofSeconds(maxDuration.asInstanceOf[Long]))
```
?

But I'd probably use milliseconds anyway, otherwise 1.9s is treated as 1s and 0.5s is treated as 0s.

Reply:
done thanks